RFC: cargo-script (#3502)
rust-lang/rfcs: text/3502-cargo-script.md

- Feature Name: `cargo-script`
- Start Date: 2023-03-31
- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
- RFC PR: [rust-lang/rfcs#3502](https://github.com/rust-lang/rfcs/pull/3502)
- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)

# Summary
[summary]: #summary

This RFC adds support for single-file bin packages in cargo.
Single-file bin packages are rust source files with an embedded manifest and a
`main`.
These files will be accepted by cargo commands as `--manifest-path` just like `Cargo.toml` files.
`cargo` will be modified to accept `cargo <file>.rs` as a shortcut to `cargo
run --manifest-path <file>.rs`;

> this allows placing `cargo` in a `#!` line for directly running these files.

Reviewer: Just realized a potentially fatal issue with using cargo like this: what if the program doesn't have an extension, and it's named `test` (not uncommon), or `bench`, or for that matter *any future name we want to use`.

Author: This is covered in the `cargo <file>.rs` section where we talk about precedence and disambiguation.

For extensionless scripts, you must have a `/` in the parameter, so `cargo ./test`.  From what I understand of shebangs being run, this will always be the case.

Reviewer: We may need a dedicated script runner program for this reason.

Author: Which cases are you concerned about?

Reviewer: On some platforms it is possible for the script to be passed as just "basename" with no ./ or other path. (I'll look further into when and how that can happen.)

Reviewer: I tested to confirm: On Linux, running a script in the current directory with `PATH` unset or set to the empty string will result in passing the basename to the interpreter with no `./` prepended.

There may be other OSes and situations in which this behavior applies.

I'm not sure if any of these scenarios are important enough to warrant a different entry point, but I think it's worth explicitly documenting this issue and confirming that we can live with it.

Author: btw I looked into how other tools handle ambiguity: https://github.com/epage/cargo-script-mvs/issues/154

dlang's dub was the easier to decipher based on the source

Author: I assume whats happening here is the CWD is being considered part of  PATH, like on Windows.  I glossed over this because my understanding is its generally frowned upon.  Its now called out in f7b8bf3e along with a couple other potential problems.

Reviewer 2: > On some platforms it is possible for the script to be passed as just "basename" with no ./ or other path. (I'll look further into when and how that can happen.)

> On Linux, running a script in the current directory with PATH unset or set to the empty string will result in passing the basename to the interpreter with no ./ prepended.

This is actually down to the shell on unix - see e.g. https://github.com/fish-shell/fish-shell/issues/9143.

Basically whatever ends up being passed to some `execXY` or other.

Author: That seems to be about absolute vs relative which I think is a reasonable thing for shells to choose one way or the other.  The proposal has no problem with that.

The problem here is when dealing with `foo` without any `/` in it.  Shells and the kernel are limited on this.  If CWD isn't in the path, they have to prepend the path for it to be able to be run.  If CWD is in the path, well, then its a problem.

Reviewer 2: I believe it is possible for it to not have a `/`, even if $PATH is not empty. If you'll forgive me, my demonstration is C (and I'm not actually any good at C):

In test.c, compile with `gcc test.c -o test`:

```C
#include <unistd.h>
#include <stddef.h>
  
int main(int argc, char **argv) {
  char *const args[2] = {"HAHAHA", NULL};
  char **env;
  extern char **environ;
 
  execve("foo", args, env);
}
```

"nameprinter.c", compile as `gcc nameprinter.c -o nameprinter`:

```C
#include <stdio.h>
  
int main(int argc, char **argv) {
    for (int i = 0; i < argc; i++) {
        printf("%d: %s\n", i, argv[i]);
    }
    return 0;
}
```

And a file called "foo", where the contents are

```
#!/absolute/path/to/nameprinter
```

(alas you'll have to add that yourself)

Make everything executable and run `PATH=/bin ./test`.

This gets me output of:

```
0: /home/alfa/nameprinter
1: foo
```

So nameprinter receives the argument "foo", without any `/`.

Up-to-date Archlinux machine, kernel 6.5.5, glibc 2.38.

Author: Let me check my understanding of the situation.  Someone can run into this if they invoke a script via a  `execve` because it isn't looking things up by `PATH` but is being given a path to the binary (so `./` can be assumed).

Reviewer 2: Yes, and the execve can be in anything. It could be in a shell, it could be in another program that runs the script for some reason (tho not via `system(3)` or equivalent because that would go via /bin/sh, which would need `./` - unless that /bin/sh removes the `./` for some reason), it could be in another interpreter.

The point being that there's no guarantee for the `/`.

You would use a valid `execve()` invocation to run the file, and it could end up running a cargo builtin.

Author: This is now at least noted as of ad5ab232

Author: We talked about this more in the cargo team meeting.  Overall, we feel that the chance of conflict will be very small and can be worked around, so we don't see anything immediately done about it.

We did talk a bit [AT_EXECFN](https://www.man7.org/linux/man-pages/man3/getauxval.3.html) and have put that down as an [Unresolved questions](https://github.com/epage/rfcs/blob/script/text/3502-cargo-script.md#unresolved-questions) to be worked out during stabilization for whether we should leverage it for detecting that we were run via shebang

- Use the playground
  - Can't access local resources
  - Limited in the crates supported
  - *Note:* there are alternatives to the playground that might have fewer
    restrictions but are either less well known or have additional
    complexities.
- Find a place to do `cargo new`, edit `Cargo.toml` and `main.rs` as necessary, and `cargo run` it, then delete it
  - This is a lot of extra steps, increasing the friction to trying things out
  - This will fail if you create in a place that `cargo` will think it should be a workspace member

By having a single-file package,
- It is easier to setup and tear down these experiments, making it more likely to happen
- All crates will be available
- Local resources are available

**One-Off Utilities:**

It is fairly trivial to create a bunch of single-file bash or python scripts
into a directory and add it to the path.
Compare this to rust where
- `cargo new` each of the "scripts" into individual directories
- Create wrappers for each so you can access it in your path, passing `--manifest-path` to `cargo run`

**Non-Goals:**

With that said, this doesn't have to completely handle every use case for

> Collaboration, Interoperability, Prototuping, or One-off Utilities.

Reviewer: ```suggestion
Collaboration, Interoperability, Prototyping, or One-off Utilities.
```

$ chmod +x hello_world.rs
$ ./hello_world.rs
Hello, world!
```

### Dependencies

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/dependencies.html))*

[crates.io] is the Rust community's central [*package registry*][def-package-registry]
that serves as a location to discover and download
[packages][def-package]. `cargo` is configured to use it by default to find
requested packages.

#### Adding a dependency

To depend on a library hosted on [crates.io], you modify `hello_world.rs`:
````rust
#!/usr/bin/env cargo
```cargo
[dependencies]
time = "0.1.12"
```

fn main() {
    println!("Hello, world!");
}
````

The data inside the `cargo` frontmatter is called a
[***manifest***][def-manifest], and it contains all of the metadata that Cargo
needs to compile your package.
This is written in the [TOML] format (pronounced /tɑməl/).

`time = "0.1.12"` is the name of the [crate][def-crate] and a [SemVer] version
requirement. The [specifying
dependencies](https://doc.rust-lang.org/cargo/guide/../reference/specifying-dependencies.html) docs have more
information about the options you have here.

If we also wanted to add a dependency on the `regex` crate, we would not need

> to add `[dependencies]` for each crate listed. Here's what your whole

Reviewer 3: "we would not need to add `[dependencies]` for each crate listed" seems like a non sequitur here (and in the source text) — what is this list to be ignored?

Author: I think this is saying that you don't do
```toml
[dependencies]
time = "0.1.12"
[dependencies]
regex = "0.1.41"
```

Reviewer 3: Ahhh, that makes sense — I didn't see the interpretation of referring to the literal text `[dependencies]`. (It would be clearer if it said "to add the `[dependencies]` section header". But that is not really about this RFC.)

`hello_world.rs` file would look like with dependencies on the `time` and `regex`
crates:

````rust
#!/usr/bin/env cargo
```cargo
[dependencies]
time = "0.1.12"
regex = "0.1.41"
```

fn main() {
    let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
    println!("Did our date match? {}", re.is_match("2014-01-01"));
}
````

You can then re-run this and Cargo will fetch the new dependencies and all of their dependencies.  You can see this by passing in `--verbose`:
```console

> $ cargo eval --verbose ./hello_world.rs

Reviewer 4: `cargo eval` was an earlier proposal, AFAICT? Looks like this example needs to be updated.

[TOML]: https://toml.io/
[crates.io]: https://crates.io/
[SemVer]: https://semver.org
[def-crate]:             https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#crate             '"crate" (glossary entry)'
[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package           '"package" (glossary entry)'
[def-package-registry]:  https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package-registry  '"package-registry" (glossary entry)'
[def-manifest]:          https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#manifest          '"manifest" (glossary entry)'

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Single-file packages

In addition to today's multi-file packages (`Cargo.toml` file with other `.rs`
files),
we are adding the concept of single-file packages which may contain an
embedded manifest.
There is no required way to distinguish a single-file
`.rs` package from any other `.rs` file.

A single-file package may contain an embedded manifest.
An embedded manifest is stored using `TOML` in rust "frontmatter",

> a markdown code-fence with `cargo` at the start of the infostring at the top of

Reviewer 5: I don't like that this makes a _language feature_ specific to Cargo. It should be possible to create alternative build tools to Cargo (similar to how JS can be used with `npm`, `yarn` or `pnpm`). The info string should not be part of the _language_ specification, nor the format (TOML).

Also, the correct term from CommonMark is a "fenced code block". A code fence is only the sequence of at least three `` ` `` or `~` characters that surround a fenced code block.

This document doesn't specify if the code fence syntax should conform to the CommonMark syntax, nor does it say what a code fence is. So it is unclear if rustc will accept this code:

`````md
   ~~~~cargo
   [package]
  ~~~~~~
`````

The [CommonMark spec](https://spec.commonmark.org/0.30/#fenced-code-blocks) would parse this as a valid fenced code block, because it says that the closing code fence must have _at least_ as many characters as the opening code fence, and the indentation of the opening and closing code fences need not match (although the indentation can't exceed 4 spaces).

Honestly, I don't think supporting all these edge cases in rustc is a good idea; that's also why I think using CommonMark syntax is a bad idea to begin with.

Author: This is a conversation for #3503

the file.

Inferred / defaulted manifest fields:
- `package.name = <slugified file stem>`
- `package.version = "0.0.0"` to [call attention to this crate being used in unexpected places](https://matklad.github.io/2021/08/22/large-rust-workspaces.html#Smaller-Tips)
  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
- `package.publish = false` to avoid accidental publishes, particularly if we
  later add support for including them in a workspace.
- `package.edition = <current>` to avoid always having to add an embedded
  manifest at the cost of potentially breaking scripts on rust upgrades

>   - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.

Reviewer: The examples shown near the top of the document don't show this warning. Could we either show the warning or include an edition?

Author: 3bcb2ad6 shows the full cargo output

  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
  - Based on feedback, we can switch to "edition is required as of <future> edition"

Disallowed manifest fields:
- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`

Single-file packages maintain an out-of-line target directory by default.

> This is `$CARGO_HOME/target/<hash-of-path>`.

Reviewer 6: This means we lose the target dir when we rename or move the script. As `<hash-of-path>` is unreadable anyway how do we feel about using something else?

A files inode number (POSIX) or File ID (Windows) stays the same between renames and moves on the same filesystem. Combining inode and a filesystem description results in a system unique identifier.

This could look like `$CARGO_HOME/target/dev/sda/232389`. 

Further more to clean up target directories for deleted *cargo-scripts* we can check if a *cargo-script* still exists at that inode (the numbers are reused as files are deleted).

Reviewer 7: It's very common for editors to replace a file with a new one on each change, to get atomic updates, which would mean that every time you edit the script it gets a new target directory and has to rebuild all dependencies.

Reviewer 6: I didnt know that, thanks for bringing it up! That rules out innodes.

I would still like to be able to move a *cargo-script* around without having to recompile. Its not *that* important to me though.

The only other approach I can think of right now is to take as little of the path as possible and expand on hash collisions. I do not think the extra complexity that would lead to is worth it though...

Author: While I've not dived into cargo fingerprinting all that much, from what I understand, moving a package around will already cause a lot ot be recompiled.  I see #5931 as the way to be reducing recompiles.

Author: I called this out in 93e8b1f

A single-file package is accepted by cargo commands as a `--manifest-path`
- Files are considered to have embedded manifest if they end with `.rs` or they lack and extension and are of type file
- This allows running `cargo test --manifest-path single.rs`
- `cargo add` and `cargo remove` may not support editing embedded manifests initially
- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)

Single-file packages will not be accepted as `path` or `git` dependencies.

> The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In

Reviewer 4: This means that lockfiles, as well as build artefacts, will just be placed in a `target` dir next to the script, right? I think this is one of the most problematic parts of this proposals. Potential problems:

- People won't know that there's a lockfile and keep using very old versions of security-relevant dependencies
- Running two scripts in the same directory via Cargo will lead to one's lockfile overwriting the other one's (?)
- It seems annoying to have a visible `target` directory at all

I will leave my other thoughts on this in the future possibilities section.

the future, when workspaces are supported, that will allow a user to have a
persistent lockfile.
We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).

## `cargo <file>.rs`

`cargo` is intended for putting in the `#!` for single-file packages:
```rust
#!/usr/bin/env cargo

fn main() {
    println!("Hello world");
}
```
- Like with `cargo install`, `.cargo/config.toml` will be read based on the
  scripts location rather than the current-dir.
  - And like `cargo install`, the current-dir rustup-toolchain is respected

> - `--release` is not passed in because the primary use case is for exploratory

Reviewer 3: Some ideas for how this situation can be improved:
* To address this in the future, there could be a manifest field that allows specifying a default profile used by `cargo run` in place of the `dev` profile. (This is orthogonal to single-file packages.)
* Today, a single-file-package author will be able to alter the `dev` profile to be optimized or otherwise similar to `release`, if they consider it a priority.

Reviewer: I think this is worth considering carefully. People will use this for quick scripts, but the fact that they're writing the scripts in *rust* may be for (among many other reasons) performance. So we should have a *very simple* way to switch to release mode.

Author: This is at least called out now in f78441ff and 9002faf4.f we do add a `run` profile, just doing

If we add the suggested `run` profile, we could ensure the following works
```toml
[profile.run]
inherits = "release"
```
That seems like a relatively light way to switch it

Reviewer 8: It would be nice if we could read arguments from the shebang, e.g.

```
#!/usr/bin/env cargo --release
```

but I imagine there are a good number of cross-platform wrinkles here.

Author: Yes, one of our items under "Naming: Considerations" is

> The command that is run in a `#!` line should not require arguments (e.g. not
> `#!/usr/bin/env cargo <something>`) because it will fail.  `env` treats the
> rest of the line as the bin name, spaces included.  You need to use `env -S`
> but that wasn't supported on macOS at least, last I tested.

Author: Just noticed we have some weird profile special cases.  I wonder if it'd be acceptable to add one more: if `run` does not have a defined table, you get the behavior of `dev`, including the directory name.

Author: We talked about this in the cargo team meeting.  We felt that there are viable, non-breaking ways of addressing this later so we decided to defer address this as it isn't the primary focus of this first phase of this work.

This is covered in https://github.com/epage/rfcs/blob/script/text/3502-cargo-script.md#access-to---release-in-shebang

  programming, so the emphasis will be on build-time performance and debugging,
  rather than runtime performance

Most other flags and behavior will be similar to `cargo run`.

The precedence for `cargo foo` will change from:
1. built-in commands
2. user aliases
3. third-party commands

to:
1. built-in command xor manifest
2. user aliases
3. third-party commands

To allow the xor, we enforce that
- manifests must be passed in as `Cargo.toml`, with a `.rs` extension, or have a `/` in them
  - So `./build` can be used to run a script name `build` rather than the `cargo build` command
- no built-in command may look like an accepted manifest

When piping `cargo <file>.rs`, `--quiet` will be assumed.
Further work may be done to refine the output in interactive mode.

# Drawbacks
[drawbacks]: #drawbacks

The implicit content of the manifest will be unclear for users.
We can patch over this as best we can in documentation but the result won't be
ideal.

> A user can workaround this with `cargo metadata --manifest-path <file>.rs`.

Reviewer 3: Idea for discoverability: Make `cargo --verbose <file>.rs` print the implicit manifest along with running the script. "Explain the steps we are taking to interpret the input" is in line with what people expect from `--verbose`.

Author: The challenge with `--verbose` (or log levels) is everyone (and every situation) has a different idea of what extra output is needed.

I feel like `cargo metadata` and `cargo read-manifest` are likely sufficient.

- A REPL is a lot larger of a problem, needing to pull in a lot of interactive behavior that is unrelated to `.rs` files
- A REPL for Rust is a lot more nebulous of a future possibility, making it pre-mature to design for it in mind

Therefore, this RFC proposes we limit the scope of the new command to `cargo run` for single-file rust packages.

## Naming
[naming]: #naming

Considerations:
- The name should tie it back to `cargo` to convey that relationship
- The command that is run in a `#!` line should not require arguments (e.g. not
  `#!/usr/bin/env cargo <something>`) because it will fail.  `env` treats the
  rest of the line as the bin name, spaces included.  You need to use `env -S`
  but that wasn't supported on macOS at least, last I tested.
- Either don't have a name that looks like a cargo-plugin (e.g. not
  `cargo-<something>`) to avoid confusion or make it work (by default, `cargo
  something` translates to `cargo-something something` which would be ambiguous
  of whether `something` is a script or subcommand)

Candidates
- `cargo-script`:
  - Out of scope
  - Verb preferred

> - `cargo-shell`:

Reviewer: ```suggestion
- `rust`:
  - Would fit well for Rust scripting
  - Would not make it clear that cargo is involved.
- `cargo-shell`:
```

Author: Added in 748beb55

require less information, rather than more.

Note: this is a reversible decision on an edition boundary

> Disposition: Rejected *for now* due to the extra boilerplate
> for throwaway scripts and not following out pattern of how we are handling
> manifests differently than `Cargo.toml`.  We might switch to this in the
> future if we find that the "latest as default" doesn't work as well as we
> expected.0

**Alternative 2: `cargo-<edition>-edition` variants**

```rust
#!/usr/bin/env cargo-2018-edition

fn main() {
}
```
single-file packages will fail if used by `cargo-<edition>` and `package.edition` are both specified.
This still needs a decision for when neither is specified.

On unix-like systems, these could be links to `cargo` can
parse `argv[0]` to extract the `edition`.

However, on Windows the best we can do is a proxy to redirect to `cargo`.

Over the next 40 years, we'll have dozen editions which will bloat the
directory, both in terms of the number of files (which can slow things down)
and in terms of file size on Windows.

This might also make shell completion of `cargo` noiser than what we have today with third-part plugins.

> > Dispition: Deferred and we'll re-evvaluate based on feedback

Reviewer: ```suggestion
> Disposition: Deferred and we'll re-evaluate based on feedback.
```

Author: Fixed in c55c10bd

  - `gorun <file>` to build and run a file
  - Implicit garbage collection for build cache
  - Project metadata is specified in HEREDOCs in regular code comments

Perl
- [Re-interprets the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops)

Ruby
- [`bundler/inline`](https://bundler.io/guides/bundler_in_a_single_file_ruby_script.html)
  - Uses a code-block to define dependencies, making them available for use

Cross-language
- [`scriptisto`](https://github.com/igor-petruk/scriptisto)
  - Supports any compiled language
  - Comment-directives give build commands
- [nix-script](https://github.com/BrianHicks/nix-script)
  - Nix version of scriptisto, letting you use any Nix dependency

See also [Single-file scripts that download their dependencies](https://dbohdan.com/scripts-with-dependencies)

# Unresolved questions
[unresolved-questions]: #unresolved-questions

# Future possibilities
[future-possibilities]: #future-possibilities

Note: we are assuming the following are **not** future possibilities in this design
- Embedding build scripts
- Embedding `.cargo/config.toml` files

> - Embedding `rustup-toolchain.toml` files

Reviewer 9: nit
```suggestion
- Embedding `rust-toolchain.toml` files
```

## Workspace Support

Allow scripts to be members of a workspace.

The assumption is that this will be opt-in, rather than implicit, so you can
easily drop one of these scripts anywhere without it failing because the
workspace root and the script don't agree on workspace membership.  To do this,
we'd expand `package.workspace` to also be a `bool` to control whether a
workspace lookup is disallowed or whether to auto-detect the workspace
- For `Cargo.toml`, `package.workspace = true` is the default
- For single-file packages, `package.workspace = false` is the default

When a workspace is specified
- Use its target directory
- Use its lock file
- Be treated as any other workspace member for `cargo <cmd> --workspace`
- Check what `workspace.package` fields exist and automatically apply them over default manifest fields
- Explicitly require `workspace.dependencies` to be inherited
  - I would be tempted to auto-inherit them but then `cargo rm`s gc will remove them because there is no way to know they are in use
- Apply all `profile` and `patch` settings

This could serve as an alternative to
[`cargo xtask`](https://github.com/matklad/cargo-xtask) with scripts sharing

> the lockfile and `target/` directory.

Reviewer 4: FWIW I think `cargo xtask.rs` could become an alternative to the current xtask pattern even without workspace support. It would allow moving `.cargo` back into `.gitignore` for some projects, which might make it worth it to ignore potential downsides... I'm not even sure what those would be, maybe the fact that `cargo xtask` can be run from arbitrary subdirectories of the workspace with the current `--package` alias setup; I don't think the same expectation applies to `cargo xtask.rs`.

<expr>` flag.

Ideally, this repl would also allow the equivalent of `python -i <file>`, not
to run existing code but to make a specific file's API items available for use
to do interactive whitebox testing of private code within a larger project.

## Embedded or adjacent Lockfile

[Lockfiles](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)
record the exact version used for every possible dependency to ensure
reproducibility.  In particular, this protects against upgrading to broken
versions and allows continued use of a yanked version.  As this time, the
recommendation is for
[`bin`s to persist their lockfile while `lib`s do not](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).

With multi-file packages, `cargo` writes a `Cargo.lock` file to the package
directory.  As there is no package directory for single-file packages, we need
to decide how to handle locking dependencies.

Considerations
- Sharing of single-file projects should be easy
  - In "copy/paste" scenarios, like reproduction cases in issues, how often
    have lockfiles been pertinent for reproduction?

> - There is an expectation of a reproducible Rust experience

Reviewer 4: I'm not sure if this is true for scripts where there is no *immediately visible* lockfile.

Author: I assume what you are looking at and saying here is that we aren't getting the reproducible experience with a hidden way lockfile?

If so, then you are highlighting what this is highlighting.  This is meant to be a list of all consideratons.  I've made some small tweaks in 967a4941

Reviewer 4: Yes? I was saying that reproducibility is maybe not expected when no lockfile is (visibly) generated, because people are already used to not getting a reproducible build (compared to other people's builds) when checking out a repo of a Rust package without a `Cargo.lock`.

Author: Thanks, that helps clarify where you are coming from.

This was written independent of some future state where cargo script, as currently state in the RFC, has been in use for a while and built up an expectation around hidden-away lockfiles.  This is speaking of the Rust experience broadly and in the present.

- Dropping of additional files might be frustrating for users to deal with (in
  addition to making it harder to share it all)
- We would need a way to store the lockfile for `stdin` without conflicting

>   with parallel runs

Reviewer 4: IMHO, it makes no sense to attempt to persist a lockfile for something as ephemeral as text from `stdin`.

Author: imo this isn't worth designing out at this point as its a future possibility interacting with a future possibility.

ac2540a8 tries to clarify that stdin is a future possibility.

  workspace at which point they would use the workspace's `Cargo.lock` file.
  This limits the scope of the conversation and allows an alternative to
  whatever is decided here.
- Read-only single-file packages (e.g. running `/usr/bin/package.rs` without root privileges)

> Disposition: Deferred.  We feel this can be handled later, either by checking
> for a manifest field, like `workspace.lock`, or by checking if the lock
> content exists (wherever it is stored).  The main constraint is that if we
> want to embed the lock content in the `.rs` file, we leave syntactic room for
> it.

**Location 1: In `CARGO_TARGET_DIR`**

The path would include a hash of the manifest to avoid conflicts.

- Transient location, lost with a `cargo clean --manifest-path foo.rs`
- Hard to find for sharing on issues, if needed

**Location 2: In `$CARGO_HOME`**

The path would include a hash of the manifest to avoid conflicts.

- Transient location though not lost with `cargo clean --manifest-path foo.rs`
- No garbage collection to help with temporary source files, especially `stdin`

**Location 3: As `<file-stem>.lock`**

Next to `<file-stem>.rs`, we drop a `<file-stem>.lock` file.   We could add a

> `_` or `.` prefix to distinguish this from the regular files in the directory.

Reviewer 4: Why drop the extension? `script.rs.lock` seems orders of magnitue less likely to conflict with regular files, shows up *after* `script.rs` in file listings and has a more obvious connection to `script.rs`.

Author: As this is in the future possibilities section, the point is more to illustrate an idea.  We don't need to design this out at this time.

- Feature Name: `cargo-script`
- Start Date: 2023-03-31
- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
- RFC PR: [rust-lang/rfcs#3502](https://github.com/rust-lang/rfcs/pull/3502)
- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)

# Summary
[summary]: #summary

This RFC adds support for single-file bin packages in cargo.
Single-file bin packages are `.rs` files with an embedded manifest and a
`main`.
These files will be accepted by cargo commands as `--manifest-path` just like `Cargo.toml` files.

> `cargo` will be modified to accept `cargo <file>.rs` as a shortcut to `cargo

Reviewer 7: Given the example doesn't use the `.rs` extension that should probably be denoted as optional here.

Author: This was meant to focus on the semantics of `.rs` files and not the literal extension, leaving discussion of that for later.  I've updated it to just say "rust source files" to avoid people reading too much into it.

run --manifest-path <file>.rs`;
this allows placing `cargo` in a `#!` line for directly running these files.

Support for single-file lib packages, publishing, and workspace support is
deferred out.

Example:
````rust
#!/usr/bin/env cargo

> ```cargo

Reviewer 10: `cargo` has been specified in the shebang above already. Why does the user need to specify it twice? It could just allow a naked ``` fence instead (at least if a shebang is present). We know that we are in a cargo-script file in that instance, so we have the freedom to set a "default" case for what to do with fences that have no specifier.

IMO defaulting them to the manifest does not paint us in a corner while saving us character overhead.

Reviewer: Specifying it leaves us much more room for future expansion, such as embedding other associated files or serving as frontmatter for other purposes. And it will make the job of syntax highlighters much easier.

If we want less appearance of redundancy, at the expense of more characters, we could use `manifest` or similar. But that doesn't seem worth it.

Reviewer 10: The start choice tends to stick around. In response to my suggestion in https://github.com/rust-lang/cargo/pull/12301 , I have concerns that it's an unfamiliar way of specifying the edition. compare it to the PR from when the edition mechanism was introduced, there one of the comments against was for a similar reason to the one you reject `cargo` defaulting now: that we don't know if we want to make `2018-pre` a thing or not. Not that I agree with them but a similar effect might occur if we require the cargo specifier to be present.

Regarding whether it should be discussed in #3503 or not, IMO that RFC should not be as restricted to precisely what cargo needs, but be more generic. Ideally when we add support for non manifest embeddings, we shouldn't have to make another lang RFC but handle it wholly inside cargo-script.

Author: > Ideally when we add support for non manifest embeddings, we shouldn't have to make another lang RFC

i doubt we'd need an RFC; likely a change proposal would do.

Author: #3503 was merged which has us make the infostring optional

files),
we are adding the concept of single-file packages which may contain an
embedded manifest.
There is no required way to distinguish a single-file
`.rs` package from any other `.rs` file.

A single-file package may contain an embedded manifest.
An embedded manifest is stored using `TOML` in rust "frontmatter",
a markdown code-fence with `cargo` at the start of the infostring at the top of
the file.

Inferred / defaulted manifest fields:
- `package.name = <slugified file stem>`
- `package.version = "0.0.0"` to [call attention to this crate being used in unexpected places](https://matklad.github.io/2021/08/22/large-rust-workspaces.html#Smaller-Tips)
  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
- `package.publish = false` to avoid accidental publishes, particularly if we
  later add support for including them in a workspace.
- `package.edition = <current>` to avoid always having to add an embedded
  manifest at the cost of potentially breaking scripts on rust upgrades
  - Warn when `edition` is unspecified.  While with single-file packages this will be

>     silenced by default, users wanting stability are also likely to be using

Author: The `While with single-file ...` section is a leftover from a previous iteration of this proposal and was removed in dd8a334a.

The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
the future, when workspaces are supported, that will allow a user to have a
persistent lockfile.
We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).

## `cargo <file>.rs`

`cargo` is intended for putting in the `#!` for single-file packages:
```rust
#!/usr/bin/env cargo

fn main() {
    println!("Hello world");
}
```
- Like with `cargo install`, `.cargo/config.toml` will be read based on the
  scripts location rather than the current-dir.
  - And like `cargo install`, the current-dir rustup-toolchain is respected
- `--release` is not passed in because the primary use case is for exploratory
  programming, so the emphasis will be on build-time performance and debugging,
  rather than runtime performance

Most other flags and behavior will be similar to `cargo run`.

The precedence for `cargo foo` will change from:
1. built-in commands
2. user aliases
3. third-party commands

to:
1. built-in command xor manifest
2. user aliases
3. third-party commands

To allow the xor, we enforce that

> - manifests must be passed in as `Cargo.toml`, `foo.rs`, or have a `/` in them

Reviewer 7: And may not have a `.` in the filename unless they end in `.rs`:

```console
> cat test.hello
#!/usr/bin/env -S cargo -Zscript

fn main() {
    println!("hello world");
}
> ./test.hello
error: the manifest-path must be a path to a Cargo.toml file
```

(assuming the current implementation has the intended semantics here)

Author: It says it must be `foo.rs`.  I've clarified that by shifting from an example to saying "with a `.rs` extension".

Reviewer 7: It still reads to me as if ‘./hello.test‘ would be allowed because it matches the third clause, but it's not because only extensionless filenames are checked for ‘/‘.

Reviewer 11: I don't personally use an iPad so I can't verify that it's impossible but last I checked if you hold `'` for a second or so it'll pop up with alternative quote options.

It'll look like so:
![Quote options on Apple Keyboard](https://github.com/rust-lang/rfcs/assets/109059814/d3a37ce4-da1f-4e03-aa10-e0c820935fcb)

Author: I see. This is specifically referring to disambiguating `cargo foo`.

When not on mobile, ill need to check if there is another place where we talk about how to parse the file.

Author: a4763f18 updated the section on this matching the current implementation in cargo.

- Feature Name: `cargo-script`
- Start Date: 2023-03-31
- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
- RFC PR: [rust-lang/rfcs#3502](https://github.com/rust-lang/rfcs/pull/3502)
- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)

# Summary
[summary]: #summary

This RFC adds support for single-file bin packages in cargo.
Single-file bin packages are `.rs` files with an embedded manifest and a
`main`.

> These will be accepted with just like `Cargo.toml` files with

Author: @ggggggggg 
> "These will be accepted with just like Cargo.toml files" doesn't read right.

I've moved this thread to the context to make it easier to follow.

Could you clarify why you feel this isn't right?

Reviewer 12: "This RFC adds support for single-file bin packages in cargo. Single-file bin packages are .rs files with an embedded manifest and a main. cargo will be modified to accept cargo <file>.rs as a shortcut to cargo run --manifest-path <file>.rs; this allows placing cargo in a #! line for directly running these files."

Above I've deleted the 2nd sentence. The paragraph communicates clearly to me the goal. I have no idea what the 2nd sentence is trying to add, but specifically the phrase "with just like X" makes no sense.

Author: I've pushed a change that tries to clarify that sentence:

> These files will be accepted by cargo commands as `--manifest-path` just like `Cargo.toml` files.

Removing things isn't always the best solution when they aren't clear.  This isn't focused on some kind of cargo internals but core aspects of how you interact with it.  This is saying that these files are interchangeable with `Cargo.toml` and you can run commands like `cargo test --manifest-path foo.rs`.

  - This is a lot of extra steps, increasing the friction to trying things out
  - This will fail if you create in a place that `cargo` will think it should be a workspace member

By having a single-file package,
- It is easier to setup and tear down these experiments, making it more likely to happen
- All crates will be available
- Local resources are available

**One-Off Utilities:**

It is fairly trivial to create a bunch of single-file bash or python scripts
into a directory and add it to the path.
Compare this to rust where
- `cargo new` each of the "scripts" into individual directories
- Create wrappers for each so you can access it in your path, passing `--manifest-path` to `cargo run`

**Non-Goals:**

With that said, this doesn't have to completely handle every use case for
Collaboration, Interoperability, Prototyping, or One-off Utilities.
Users can always scale up to normal packages with an explicit `Cargo.toml` file.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### Creating a New Package

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/creating-a-new-project.html))*

> To start a new [package][def-package] with Cargo, create a file named `hello_world.rs`:

Author: In other places, I refer to it as a "single file package" and within the context of cargo concepts and literature, these are a form of a package.  While we might find places we want to use a term like "script" for these (which I personally am hesitant on, only using it for historical connection to `cargo-script`), we shying away from calling these "packages" could end up making things more confusing.

Granted, this gets into whether a looser or more strict term is appropriate when we finally document this and I think that is something to work out over time and likely will evolve.

}
````

You can then re-run this and Cargo will fetch the new dependencies and all of their dependencies.  You can see this by passing in `--verbose`:
```console
$ cargo --verbose ./hello_world.rs
warning: `package.edition` is unspecified, defaulting to `2021`
      Updating crates.io index
   Downloading memchr v0.1.5
   Downloading libc v0.1.10
   Downloading regex-syntax v0.2.1
   Downloading memchr v0.1.5
   Downloading aho-corasick v0.3.0
   Downloading regex v0.1.41
     Compiling memchr v0.1.5
     Compiling libc v0.1.10
     Compiling regex-syntax v0.2.1
     Compiling memchr v0.1.5
     Compiling aho-corasick v0.3.0
     Compiling regex v0.1.41
     Compiling hello_world v0.1.0 (file:///path/to/package/hello_world)
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `/home/epage/.cargo/target/98/07dcd6510bdcec/debug/hello_world`
Did our date match? true
```

## Package Layout

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/project-layout.html))*

When a single file is not enough, you can separately define a `Cargo.toml` file along with the `src/main.rs` file.  Run
```console
$ cargo new hello_world --bin

> ```

Reviewer 13: I guess this is what is mentioned in the "Scaling up" section below. Feel free to resolve the thread :)

The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
the future, when workspaces are supported, that will allow a user to have a
persistent lockfile.
We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).

## `cargo <file>.rs`

`cargo` is intended for putting in the `#!` for single-file packages:
```rust
#!/usr/bin/env cargo

fn main() {
    println!("Hello world");
}
```
- Like with `cargo install`, `.cargo/config.toml` will be read based on the
  scripts location rather than the current-dir.
  - And like `cargo install`, the current-dir `rust-toolchain.toml` is respected
- `--release` is not passed in because the primary use case is for exploratory
  programming, so the emphasis will be on build-time performance and debugging,
  rather than runtime performance

Most other flags and behavior will be similar to `cargo run`.

The precedence for `cargo foo` will change from:
1. built-in commands
2. user aliases
3. third-party commands

to:
1. built-in command xor manifest
2. user aliases
3. third-party commands

To allow the xor, we enforce that

> - manifests must be passed in as `Cargo.toml`, with a `.rs` extension, or have a `/` in them

Author: > manifests must be passed in as (`Cargo.toml` | with a `.rs` extension | have a `/` in them)

English...

New wording

> we only assume the argument is a manifest if it has a `/` in it, ends with the `.rs` extension, or is `Cargo.toml`

(this also resolves another point of confusion someone had: that this was dictating the rules generally for single-file packages and not jut for `cargo something`)

> - manifests must be passed in as `Cargo.toml`, with a `.rs` extension, or have a `/` in them

Reviewer 13: ohhhh, that makes a lot more sense. thanks for the clarification :)

  - So `./build` can be used to run a script name `build` rather than the `cargo build` command
- no built-in command may look like an accepted manifest

When piping `cargo <file>.rs`, `--quiet` will be assumed.
Further work may be done to refine the output in interactive mode.

# Drawbacks
[drawbacks]: #drawbacks

The implicit content of the manifest will be unclear for users.
We can patch over this as best we can in documentation but the result won't be
ideal.
A user can workaround this with `cargo metadata --manifest-path <file>.rs`
or `cargo read-manifest --manifest-path <file>.rs`

Users may be using this feature for heavier duty scripting and want optimized builds.
- This can be worked around by changing `profile.dev` in the embedded manifest

Like with all cargo packages, the `target/` directory grows unbounded.
This is made worse by them being out of the way and the scripts are likely to be short-lived.
Some prior art include a cache GC but that is also to clean up the temp files
stored in other locations
(our temp files are inside the `target/` dir and should be rarer).

> A GC for cargo is being tracked in [rust-lang/cargo#12633](https://github.com/rust-lang/cargo/issues/12633)

Reviewer 13: it looks like the "Dealing with leaked `target/`" section confirms that `cargo clean` will work, but it might be worth explicitly calling this out here :)

> A GC for cargo is being tracked in [rust-lang/cargo#12633](https://github.com/rust-lang/cargo/issues/12633)

Author: Mentioned in 99d5ea32

- `cargo expand` and other third-party cargo commands
- `rust-analyzer` and other editor/IDE integration

While other third-party cargo commands might not immediately adopt single-file
packages, first-party support for them will help encourage their adoption.

This still leaves room for third-party implementations, either differentiating themselves or experimenting with
- Alternative caching mechanisms for lower overhead
- Support for implicit `main`, like doc-comment examples
- Template support for implicit `main` for customizing `use`, `extern`, `#[feature]`, etc
- Short-hand dependency syntax (e.g. `//# serde_json = "*"`)
- Prioritizing other workflows, like runtime performance

## File association on Windows

We could add a non-default association to run the file.
We don't want it to be a default,
to avoid unintended harm and due to the likelihood someone is going to want to
edit these files.

## File extension

Should these files use `.rs` or a custom file extension?

Reasons for a unique file type
- Semantics are different than a normal `.rs` file
  - Except already a normal `.rs` file has context-dependent semantics (`build.rs`, `main.rs`, `random.rs`),

>     so this doesn't seem too far off

Author: Whether we used a new syntax, doc comments, or something else, its all the same.  Its rust syntax that is only understood within the context for which its used.

>     so this doesn't seem too far off

Reviewer 13: hmm, I'm not sure I agree with that statement.

in some other part of the RFCs it is mentioned that these scripts could potentially live anywhere, including `src/`, which means they could be used as regular Rust modules, when ignoring the embedded manifest.

- when using a comment-based frontmatter the parser would still understand the regular Rust code in the file and only need adjustments if it wants to understand the additional manifest.
- with the backticks the parser would likely throw up immediately and not parse the rest of the file at all.

>     so this doesn't seem too far off

Author: > with the backticks the parser would likely throw up immediately and not parse the rest of the file at all.

Isn't this dependent on whether we support frontmatters only one root mod's or on any mod?   And the decision on that is tied to the decision on allowing this in regular files.

- Different file associations for Windows
- Better detection by tools for the new semantics (particularly `rust-analyzer`)

Downsides to a custom extension

> - Limited support by different tools (rust-analyzer, syntax highlighting, non-LSP editor actions) as adoptin rolls out

Reviewer 13: ```suggestion
- Limited support by different tools (rust-analyzer, syntax highlighting, non-LSP editor actions) as adoption rolls out
```

Reviewer 13: IMHO the same is also true if we use `.rs`, since all of those tools currently don't know how to handle the frontmatter part of the script file

Author: Parsers and tools tend to be forgiving of syntax.

Author: Also, the frontmatter is optional while an extension wouldn't be

the downsides to a slow roll out.

For Windows, a different file extension doesn't buy us all that much.
We could have a "run" action associated with the extension when clicking on the
file but the most likely action people would want is to edit, not run, and
there might be concern over running code unexpectedly.
More interesting is the commandline but we do not know of a accepted equivalent of `#!` for `cmd`.
Generally, users just reference the interpreter (`python x.py`) or add a `x.bat` wrapper.

While `rust-analyzer` needs to be able to distinguish regular `.rs` files from
single-file packages to look up the relevant manifest to perform operations, we
propose that be through checking the `#!` line (e.g.
[how perl detects perl in the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops).
While this adds boilerplate for Windows developers, this helps encourage
cross-platform development.

If we adopted a unique file extensions, some options include:
- `.crs` (used by `cargo-script`)
- `.ers` (used by `rust-script`)
  - No connection back to cargo
- `.rss`
  - No connection back to cargo

>   - Confused with RSS

Reviewer 13: ```suggestion
  - Confused with [RSS](https://en.wikipedia.org/wiki/RSS)
```

😉

- If we add additional frontmatter blocks in the future (like for lockfiles),
  the manifest block will be needed at minimum and other blocks will likely
  only be used in 1% of cases the manifest block is present, so we shouldn't
  hurt the simple, common case for the more advanced, long tail cases.

## `edition`

[The `edition` field controls what variant of cargo and the Rust language to use to interpret everything.](https://doc.rust-lang.org/edition-guide/introduction.html)

A policy on this needs to balance
- Matching the expectation of a reproducible Rust experience
- Users wanting the latest experience, in general
- Boilerplate runs counter to experimentation and prototyping, particularly in the "no dependencies" case
  - A `cargo new --script` (flag TBD) could help reduce writing of boilerplate.
- There might not be a backing file if we read from `stdin` (future possibility)

**Solution: Latest as Default**

Default to the `edition` for the current `cargo` version, assuming single-file
packages will be transient in nature and users will want the current `edition`.
However, we will produce a warning when no `edition` is specified, nudging

> people towards reproducible code.

Reviewer 13: I'm wondering about the term "warning" here. This seems more like an informational thing to me. If we use "warning" then the user may feel like they did something wrong and needs to fix it, even if it runs just fine for the prototyping work they are doing.

Reviewer 13: unrelated to the above, is the compiler clever enough to tell if a compilation error happened because the wrong edition was used? if that was the case we could potentially only show the warning if something actually fails

Author: We are walking a fine line here and I think "warning" level is appropriate.  Once we have rust-lang/cargo#12235, "diagnostic" might be a more appropriate word which could be fixed with `cargo fix`.

We are trying to balance
- Low boilerplate (i.e. don't make the frontmatter required)
- Behave as the user would expect (i.e. the regular 2015 default would likely be confusing and would just force people into using a frontmatter)
- Maintain the aims of the edition system, particularly that code doesn't break

In a lot of the previous discussions, some have voiced concerns that a warning might not be enough and I'm surprised it hasn't come up here yet (for any who feel this way, I'd recommend trying it out!).

Warnings are fine for prototyping (there is a reason people generally recommend against `#![deny(warnings)]` but instead recommend `RUSTFLAGS=-Dwarnings`) but the fact that we might break people's code in upgrade is something they really need to be aware of and cautious about, hence a warning.

- Can run any file in a project (they can have their own "main") to do whitebox exploratory programming and not just blackblox

Go
- [`gorun`](https://github.com/erning/gorun/) attempts to bring that experience to a compiled language, go in this case
  - `gorun <file>` to build and run a file
  - Implicit garbage collection for build cache
  - Project metadata is specified in HEREDOCs in regular code comments

Perl
- [Re-interprets the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops)

Ruby
- [`bundler/inline`](https://bundler.io/guides/bundler_in_a_single_file_ruby_script.html)
  - Uses a code-block to define dependencies, making them available for use

Cross-language
- [`scriptisto`](https://github.com/igor-petruk/scriptisto)
  - Supports any compiled language
  - Comment-directives give build commands
- [nix-script](https://github.com/BrianHicks/nix-script)
  - Nix version of scriptisto, letting you use any Nix dependency

See also [Single-file scripts that download their dependencies](https://dbohdan.com/scripts-with-dependencies)

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> - Wherther `version` should may be unset or defaulted to `0.0.0`

Reviewer 13: ```suggestion
- Whether `version` should may be unset or defaulted to `0.0.0`
```

Reviewer 13: if we don't set it, how would things like https://docs.rs/clap/4.4.6/clap/macro.crate_version.html react to that?

Author: Thats something we'll need to work out as part of the version rust-lang/cargo#9829 if we decide to do it.

# Future possibilities
[future-possibilities]: #future-possibilities

Note: we are assuming the following are **not** future possibilities in this design
- Embedding build scripts
- Embedding `.cargo/config.toml` files
- Embedding `rust-toolchain.toml` files
- Embedding other source files or additional packages

## Playground support

The [playground](play.rust-lang.org/) could gain support for embedded
manifests, allowing users access to more packages, specific package versions,
and specific feature sets.

## Dealing with leaked `target/`

As `target/` is out of sight, it is easy to "leak" them, eating up disk space.
Users would need to know to run `cargo clean --manifest-path foo.rs` before
deleting `foo.rs` or to just do `rm -rf ~/.cargo/target` and wipe everything
(since the specific target directory will be non-obvious).

In the future we can
[track embedded manifests and garbage collect their `target/`](https://github.com/rust-lang/cargo/issues/12633).

## `cargo new` support

A `cargo new <flag> foo.rs` could generate a source file with an embedded manifest.

Questions
- Should we do this implicitly with the `.rs` extension?

> - If we have a flag, what should we name it?

Reviewer 13: I'd go with `--script`, though I'm not sure if `cargo new` really needs to support it if the goal is to be as boilerplate-free as possible

See
[t-cargo zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Re-thinking.20cargo's.20output)
and [rust-lang/cargo#8889](https://github.com/rust-lang/cargo/issues/8889).

Note: `--quiet` is inferred when piping to a file ([rust-lang/cargo#12305](https://github.com/rust-lang/cargo/pull/12305))

## Executing `<stdin>`

We could extend this to allow accepting single-file packages from stdin, either
explicitly with `-` or implicitly when `<stdin>` is not interactive.

## Implicit `main` support

Like with doc-comment examples, we could support an implicit `main`.

Ideally, this would be supported at the language level
- Ensure a unified experience across the playground, `rustdoc`, and `cargo`
- `cargo` can directly run files rather than writing to intermediate files
  - This gets brittle with top-level statements like `extern` (more historical) or bin-level attributes

Behavior can be controlled through editions

## `[lib]` support

In an effort to allow low-overhead packages in a workspace, we may also allow `[lib]`s to be defined.

A single-file package may only be a `[bin]` or a `[lib]` and not both.

> We would support depending on these, publishing them, etc.

Reviewer 13: publishing them means that the packages would be published without readme and license files and most likely other metadata though, so I'm a bit hesitant regarding this...

Author: As this is a future possibility, this is more helpful to add flavor / context for discussions and decisions.  We can work those issues out if/when we come to this.

the lockfile and `target/` directory.

## Scaling up

We provide a workflow for turning a single-file package into a multi-file
package, on `cargo-new` / `cargo-init`.  This would help smooth out the
transition when their program has outgrown being in a single-file.

## A REPL

See the [REPL exploration](https://github.com/epage/cargo-script-mvs/discussions/102)

In terms of the CLI side of this, we could name this `cargo shell` where it
drops you into an interactive shell within your current package, loading the
existing dependencies (including dev).  This would then be a natural fit to also have a `--eval
<expr>` flag.

Ideally, this repl would also allow the equivalent of `python -i <file>`, not
to run existing code but to make a specific file's API items available for use
to do interactive whitebox testing of private code within a larger project.

## Embedded or adjacent Lockfile

[Lockfiles](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)
record the exact version used for every possible dependency to ensure
reproducibility.  In particular, this protects against upgrading to broken

> versions and allows continued use of a yanked version.  As this time, the

Reviewer 13: ```suggestion
versions and allows continued use of a yanked version.  At this time, the
```

recommendation is for

> [`bin`s to persist their lockfile while `lib`s do not](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).

Reviewer 13: wasn't this changed recently? (see https://blog.rust-lang.org/2023/08/29/committing-lockfiles.html)

Author: Fixed in ad7ece27

With multi-file packages, `cargo` writes a `Cargo.lock` file to the package
directory.  As there is no package directory for single-file packages, we need
to decide how to handle locking dependencies.

Considerations
- Sharing of single-file projects should be easy
  - In "copy/paste" scenarios, like reproduction cases in issues, how often
    have lockfiles been pertinent for reproduction?
- There is an expectation of a reproducible Rust experience across systems with lockfiles being a key element
- Dropping of additional files might be frustrating for users to deal with (in
  addition to making it harder to share it all)
- We would need to decide what to do for lockfiles and `stdin` (another future possibility) without conflicting

>   with parallel runs

Reviewer 13: the current plan is apparently to use a hash of the file path, but that would also be difficult for `stdin`. would it make sense to hash the file content instead?

Author: Yes, that is a possibility  As this is a future possibility, the point is more to highlight the idea for us to include in consideration.  I don't think we need to work out all the details at this time.

[def-manifest]:          https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#manifest          '"manifest" (glossary entry)'

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Single-file packages

In addition to today's multi-file packages (`Cargo.toml` file with other `.rs`
files),
we are adding the concept of single-file packages which may contain an
embedded manifest.
There is no required way to distinguish a single-file
`.rs` package from any other `.rs` file.

A single-file package may contain an embedded manifest.
An embedded manifest is stored using `TOML` in rust "frontmatter",
a markdown code-fence with `cargo` at the start of the infostring at the top of
the file.

Inferred / defaulted manifest fields:
- `package.name = <slugified file stem>`
- `package.edition = <current>` to avoid always having to add an embedded
  manifest at the cost of potentially breaking scripts on rust upgrades
  - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.

>   - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand

Reviewer: Nit:
```suggestion
  - Based on feedback, we might add `cargo-<edition>` proxies to put in `#!` as a shorthand
```

    - All other `cargo install`s will load config from `CARGO_HOME`
  - Unlike `cargo install`, we expect people to run single-file packages in unsafe locations, like temp directories or download directories, and don't want to pick up less trustworthy configs
- Like all cargo commands, including `cargo install`, the current-dir `rust-toolchain.toml` is respected ([cargo#7312](https://github.com/rust-lang/cargo/issues/7312))
- `--release` is not passed in because the primary use case is for exploratory
  programming, so the emphasis will be on build-time performance and debugging,
  rather than runtime performance

Most other flags and behavior will be similar to `cargo run`.

The precedence for `cargo foo` will change from:
1. built-in commands
2. user aliases
3. third-party commands

to:
1. built-in command xor manifest
2. user aliases
3. third-party commands

To allow the xor, we enforce that
- we only assume the argument is a manifest if it has a `/` in it, ends with the `.rs` extension, or is `Cargo.toml`
  - So `./build` can be used to run a script name `build` rather than the `cargo build` command
- no built-in command may look like an accepted manifest

> When piping `cargo <file>.rs`, `--quiet` will be assumed.

Reviewer: ```suggestion
When the stdout or stderr of `cargo <file>.rs` is not going to a terminal, cargo will assume `--quiet`.
```

their current working directory but that doesn't work for embedded manifests.

It would be helpful if we had an alias for this argument to make it easier to
specify, whether it was a short word for a "short flag".

Options include
- `--script` with `-s` or `-S` for a short flag, but is the meaning clear
  enough?  What about in the context of multi-file packages taking advantage
  of it?
- `p` is taken by `--package`
- `-m`, `-M`, and `-P` are available, but are the meanings clear enough?

## Cleaner output

`cargo foo.rs` is just a wrapper around `cargo run --manifest-path foo.rs` and cargo can be quite noisy.
The out-of-tree prototype for this instead ran `cargo run --quiet
--manifest-path foo.rs` but that was confusing as it is unclear when a long
execution time is from a slow compile or from the program.

In the future, we could try to find ways to better control cargo's output so at
least `cargo foo.rs` is less noisy but appears responsive.

See
[t-cargo zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Re-thinking.20cargo's.20output)
and [rust-lang/cargo#8889](https://github.com/rust-lang/cargo/issues/8889).

> Note: `--quiet` is inferred when piping to a file ([rust-lang/cargo#12305](https://github.com/rust-lang/cargo/pull/12305))

Reviewer: ```suggestion
Note: `--quiet` is inferred when redirecting/piping output ([rust-lang/cargo#12305](https://github.com/rust-lang/cargo/pull/12305))
```

- This will be more awkward to copy/paste and browse in bug reports as just a `serde_json` lockfile is 89 lines long
- This makes it harder to resolve conflicts (users can't just checkout the old file and have it re-resolved)
- A policy is needed when the location is read-only
  - Fallback to a user-writeable location
  - Always re-calculate the lockfile
  - Error

**Configuration 1: Hardcoded**

Unless as a fallback due to a read-only location, the user has no control over
the lockfile location.

**Configuration 2: Command-line flag**

`cargo generate-lockfile --manifest-path <file>.rs` would be special-cased to
write the lockfile to the persistent location and otherwise we fallback to a
no-visible-lockfile solution.

- Passing flags in a `#!` doesn't work cross-platform

**Configuration 3: A new manifest field**

We could add a `workspace.lock` field to control some lockfile location
behavior, what that is depends on the location on what policies we feel
comfortable making.  This means we would allow limited access to the
`[workspace]` table (currently the whole table is verboten).

- Requires manifest design work that is likely specialized to just this feature

> **Configuration 4: Exitence Check**

Reviewer: ```suggestion
**Configuration 4: Existence Check**
```

However, at best we can drop the file to 51, 102, or 153 lines (1-3 per package) which can still be overwhelming.

**Format 2: Minimal Versions**

Instead of tracking a distinct lockfile, we can get most of the benefits with
[`-Zminimal-versions`](https://github.com/rust-lang/cargo/issues/5657).

- Consistent runs across machines without a lockfile
- More likely to share versions across single-file packages, allowing more
  reuse within the shared build cache
- Deviates from how resolution typically happens, surprising people
- Not all transitive dependencies have valid version requirements

**Format 3: Timestamp**

If we record timestamps with package publishes, we could resolve to a specific
timestamp for registry packages.

Challenges:
- Cargo preserves existing resolved versions when dealing with a new or modified
  dependency or `cargo update -p`.
- Non-registry dependenciess.

If we want this to be near-lossless, it seems like we'd need
- An edit log, rather than simple timestamps
- Regular lockfile entries for non-registry dependencies

See also
[Cargo time machine (generate lock files based on old registry state) ](https://github.com/rust-lang/cargo/issues/5221)

> 

Reviewer: ```suggestion

**Format 4: Minimal lockfile format**

We could simplify the lockfile format to store less information, making it more appealing to embed in the binary. We could store enough information to easily recreate the lockfile, without requiring timestamp-based recreation of the index state. This would primarily include the exact versions of every dependency.
```

By having a single-file package,
- It is easier to setup and tear down these experiments, making it more likely to happen
- All crates will be available
- Local resources are available

**One-Off Utilities:**

It is fairly trivial to create a bunch of single-file bash or python scripts
into a directory and add it to the path.
Compare this to rust where
- `cargo new` each of the "scripts" into individual directories
- Create wrappers for each so you can access it in your path, passing `--manifest-path` to `cargo run`

**Non-Goals:**

With that said, this doesn't have to completely handle every use case for
Collaboration, Interoperability, Prototyping, or One-off Utilities.
Users can always scale up to normal packages with an explicit `Cargo.toml` file.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

### Creating a New Package

*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/creating-a-new-project.html))*

To start a new [package][def-package] with Cargo, create a file named `hello_world.rs`:
```rust
#!/usr/bin/env cargo

fn main() {
    println!("Hello, world!");
}
```

Let's run it
```console

> $ chmod +x hello_world.rs

Author: I used it as I felt it worked well as a Guide level way of introducing this feature.  I can go either way on whether we emphasize this in the official Guide or not.

> $ chmod +x hello_world.rs

Reviewer: `chmod` will exist on any platform where `#!/usr/bin/eng cargo` works. :)

A single-file package may contain an embedded manifest.
An embedded manifest is stored using `TOML` in rust "frontmatter",
a markdown code-fence with `cargo` at the start of the infostring at the top of
the file.

Inferred / defaulted manifest fields:
- `package.name = <slugified file stem>`
- `package.edition = <current>` to avoid always having to add an embedded
  manifest at the cost of potentially breaking scripts on rust upgrades
  - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.
  - Based on feedback, we might add `cargo-<edition>` proxies to put in `#!` as a shorthand
  - Based on feedback, we can switch to "edition is required as of <future> edition"

Note: As of [rust-lang/cargo#123](https://github.com/rust-lang/cargo/pull/12786),
when `package.version` is missing,
it gets defaulted to `0.0.0` and `package.publish` gets defaulted to `false`.

Disallowed manifest fields:
- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`

> Single-file packages maintain an out-of-line target directory by default.

Author: I've added a clarifying comment that there aren't stability requirements on this.

The [`cargo-script`](https://crates.io/crates/cargo-script) family of tools has a single command for
- Run `.rs` files with embedded manifests
- Evaluate command-line arguments (`--expr`, `--loop`)

This behavior (minus embedded manifests) mirrors what you might expect from a
scripting environment, minus a REPL.  We could design this with the future possibility of a REPL.

However
- The needs of `.rs` files and REPL / CLI args are different, e.g. where they get their dependency definitions
- A REPL is a lot larger of a problem, needing to pull in a lot of interactive behavior that is unrelated to `.rs` files
- A REPL for Rust is a lot more nebulous of a future possibility, making it pre-mature to design for it in mind

Therefore, this RFC proposes we limit the scope of the new command to `cargo run` for single-file rust packages.

## Naming
[naming]: #naming

Considerations:
- The name should tie it back to `cargo` to convey that relationship
- The command that is run in a `#!` line should not require arguments (e.g. not
  `#!/usr/bin/env cargo <something>`) because it will fail.  `env` treats the
  rest of the line as the bin name, spaces included.  You need to use `env -S`

>   but that wasn't supported on macOS at least, last I tested.

Reviewer 14: A lot of decision is based around "`env -S` is not standard in particular macOS", but is it actually verified? @epage

1. On latest macOS Sonoma 14.4.1, `/usr/bin/env` *does* support an `-S` flag. [^1]
2. I have also got a macOS Big Sur 11.7.10 (2020) and its `env` also has the `-S` flag.
3. According to [Julia language's FAQ](https://docs.julialang.org/en/v1/manual/faq/#How-do-I-pass-options-to-julia-using-#!/usr/bin/env?), macOS supported `env -S` since Sierra 10.12 (2016), **and Sierra is the minimum supported macOS version [since Rust 1.74](https://blog.rust-lang.org/2023/09/25/Increasing-Apple-Version-Requirements.html)**.

Yes it is indeed still awkward to require the `-S` flag, but outright denying any alternatives that requires `-S` is not reasonable either.

[^1]: Additionally, unlike Linux, even skipping the `-S` flag it works just fine:
    ```
    #!/usr/bin/env cargo +nightly -Zscript
    fn main() { println!("works!"); }
    ```

Author: I've updated the text to not mention macOS.  I was going off of memory when I did that (it was a problem at a prior job and most problems came from macOS).  I have no idea if it was due to our Docker base images or something about macOS pre-Sierra.

I did confirm that `busybox` does not support `-S`.  The broader point is that its not portable.

## Script-relative config

As `.cargo/config.toml` is loaded from `CARGO_HOME`, there isn't a way to ensure that we load the config for a script in a repo (e.g. an xtask).
This could become more of prevalent of an issue when workspaces are supported.

Options
- A way to opt-in saying that the parent directories are assumed to be safe directories (see also [RFC #3279](https://github.com/rust-lang/rfcs/pull/3279)).
- Key off of workspace membership (as then it won't likely be a temp file)
  - This has a chicken-and-egg problem as we need to load config before we load manifests

## Access to `--release` in shebang

While the primary target of this feature is one-off use that is not performance sensitive,
users may want to have long-lived scripts that do intensive operations.

This is generally viewed as a script-author decision, rather than a user decision, as they are most likely to know the performance characteristics of their script.

In the short term, users can configure `[profile.dev]` to match `[profile.release]`, see [Cargo's Profiles chapter](https://doc.rust-lang.org/cargo/reference/profiles.html#default-profiles).

Potential options for addressing this include

> - Add `cargo --release <script>` or `cargo --profile=name> <script>`

Reviewer 14: ```suggestion
- Add `cargo --release <script>` or `cargo --profile=<name> <script>`
```
